website/docs/sources/discord/index.md

```diff
@@ -255,35 +255,44 @@ if not connection:
 access_token = connection.access_token
 
 guild_member_info = requests.get(
-    GUILD_API_URL.format(guild_id=GUILD_ID),
-    headers={"Authorization": "Bearer " + access_token},
+  GUILD_API_URL.format(guild_id=guild_id),
+  headers={
+    "Authorization": "Bearer " + access_token
+  },
 ).json()
 
 # Ensure user is a member of the guild
 if "code" in guild_member_info:
-    if guild_member_info["code"] == 10004:
+    if guild_member_info['code'] == 10004:
         ak_message("User is not a member of the guild")
     else:
-        ak_create_event(
-            "discord_error", source=context["source"], code=guild_member_info["code"]
-        )
+        ak_create_event("discord_error", source=context['source'], code=guild_member_info['code'])
         ak_message("Discord API error, try again later.")
     return False
 
-# Get the user's current roles and remove all roles we want to remap
-new_groups = [
-    role for role in request.user.ak_groups.all() if role not in MAPPED_ROLES.values()
-]
+# Get all discord_groups
+discord_groups = Group.objects.filter(attributes__discord_role_id__isnull=False)
+
+# Get all user groups except discord_groups
+user_groups = request.user.ak_groups.exclude(pk__in=discord_groups.values_list('pk', flat=True))
 
-# Add back mapped roles which the user has in the guild
-for role_id in MAPPED_ROLES:
-    if role_id in guild_member_info["roles"]:
-        new_groups.append(MAPPED_ROLES[role_id])
+# Filter matching roles based on guild_member_info['roles']
+matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info['roles'])
+
+# Combine user_groups and matching_roles
+combined_groups = user_groups.union(matching_roles)
 
 # Update user's groups
-request.user.ak_groups.set(new_groups)
+request.user.ak_groups.set(combined_groups)
 request.user.save()
 
+# Create event with roles changed
+ak_create_event(
+    "discord_role_sync",
+    discord_roles_before=', '.join(str(group) for group in request.user.ak_groups.filter(pk__in=discord_groups.values_list('pk', flat=True))),
+    discord_roles_after=', '.join(str(group) for group in matching_roles),
```

Comment:
```suggestion
    discord_roles_before=", ".join(str(group) for group in request.user.ak_groups.filter(pk__in=discord_groups.values_list("pk", flat=True))),
    discord_roles_after=", ".join(str(group) for group in matching_roles),
```

Reply:
Those queries should already be available from above. Why not reuse them instead of doing additional database queries?

Reply:
> Those queries should already be available from above. Why not reuse them instead of doing additional database queries?

Do you mean ``request.user_ak_groups`` or ``request.user.ak_groups.filter(pk__in=discord_groups.values_list('pk', flat=True))``?

If you are saying that ``request.user.ak_groups`` is calculated on every access, we can definitely do that.
If you are referring to the value of ``discord_roles_before`` then the following should explain why we can't reuse above code:

In the code preceeding the event we filter out roles which are __not__ in `discord_groups`.
However in this event, we want to display `discord_groups` the user was in.

Basically we have these relations:
`User-Roles` - `discord_groups` (preceeding code)
`User-Roles` ∩ `discord_groups` (within event)

If I am not mistaken, one cannot easily be reused for the other

Please let me know which you mean so I can adjust the code.

Reply:
So actually, the second one is fine, but the first one, `discord_roles_before` is I think wrong. As written, the query is based on `request.user.ak_groups` _after_ it's already been modified. So if any roles were removed, they wouldn't show up in the event.

Reply:
> So actually, the second one is fine, but the first one, `discord_roles_before` is I think wrong. As written, the query is based on `request.user.ak_groups` _after_ it's already been modified. So if any roles were removed, they wouldn't show up in the event.

I checked and it seems to be working fine.
I suppose the request-object is just the in-time data when the policy is executed.

Reply:
No, it will actually execute the `request.user.ak_groups.filter(pk__in=discord_groups.values_list("pk", flat=True))` query after you've modified the data. If we really want to log proper data, that query should be done before changing ak_groups, saved as a variable and used when needed.

Reply:
> No, it will actually execute the `request.user.ak_groups.filter(pk__in=discord_groups.values_list("pk", flat=True))` query after you've modified the data. If we really want to log proper data, that query should be done before changing ak_groups, saved as a variable and used when needed.

I tested it on 2024.4.3 and ``discord_roles_before`` was showing correctly. But we can change it if you prefer.

Reply:
Yeah, I'd rather we change it, so we're 100% sure it works